### ISSUE TITLE
Thread safety using WSAOVERLAPPED

### ISSUE THREAD

**Issue by maxsharabayko** [Haivision/srt#2838]
PR #2834 has a problem, The srt::CChannel::sendto will be call in CSndQueue(thread1) and CRcvQueue(thread2), and that not thread safety.

Them used same WSAOVERLAPPED, so WSASendTo maybe not completed when WSAWaitForMultipleEvents return.

From https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsasendto
> The lpOverlapped parameter must be valid for the duration of the overlapped operation. If multiple I/O operations are simultaneously outstanding, each must reference a separate WSAOVERLAPPED structure.

_Originally posted by @mGaosi in https://github.com/Haivision/srt/issues/2834#issuecomment-1870012119_

**Comment by mGaosi** [comment]
We found this issue during our app testing. When sending 400mbps of data in our program, some exceptions quickly occurred and the stack was damaged.  *The probability of this problem occurring is related to your actual traffic and the performance of the network subsystem.*

I am currently using a temporary solution, using  TLS WSAEvent and stack WSAOVERLAPPED.

```cpp
    WSAOVERLAPPED overlapped;
    ::SecureZeroMemory(&overlapped, sizeof(overlapped));
    class WSAEventRef
    {
    public:
        WSAEventRef()
            : e(::WSACreateEvent())
        {
        }
        ~WSAEventRef()
        {
            ::WSACloseEvent(e);
            e = NULL;
        }
        void reset()
        {
            ::WSAResetEvent(e);
        }
        WSAEVENT handle()
        {
            return e;
        }

    private:
        WSAEVENT e;
    };
    thread_local WSAEventRef lEvent;
    overlapped.hEvent = lEvent.handle();
    int res = ::WSASendTo(m_iSocket, (LPWSABUF)packet.m_PacketVector, 2, &size, 0, addr.get(), addrsize, &overlapped, nullptr);
```

**Comment by maxsharabayko** [comment]
The Windows build of SRT enables C++11 by default. If disabled, the pthread port is being used. But it is waiting to be deprecated.

@mGaosi 
I believe your solution works well and is worth to make a PR out of it. I would just do something like this:
```c++
#if HAVE_CXX11
thread_local WSAEventRef lEvent;
#else
WSAEventRef lEvent;
#endif
```

P.S. Look like `HAVE_CXX11` and `ENABLE_CXX11` are not working together. 🤔

**Comment by ethouris** [comment]
Without C++11, and this way using POSIX thread API, it should rather use `pthread_once`.

**Comment by ethouris** [comment]
If we state that Windows can only be built with C++11, then it should suffice to declare `thread_local WSAEventRef lEvent` under a condition for WIN32 only.

**Comment by maxsharabayko** [comment]
Resolved by #2929.
